### ISSUE TITLE
Terraform console swallows stdout when receiving input from a named pipe

### ISSUE THREAD

**Issue by BatmanAndJoker123** [hashicorp/terraform#32402]
### Terraform Version

```shell
I have a terraform file:


terraform {
  required_version = "1.3.5"
}


locals {
  a = "foo"
  b = "bar"
}


in a bash terminal, I can do:

```bash
$ echo "local.a" | terraform console
"foo"
$ echo "local.b" | terraform console
"bar"
```

Now what I'm trying to do is start a process running `terraform console` in the background and feed it commands.

This is what I've tried (following this answer https://serverfault.com/a/815253):

```bash
$ mkfifo /tmp/srv-input
$ tail -f /tmp/srv-input | terraform console >>output.txt 2>&1 &
```

this starts the background process correctly:

```bash
$ ps -ax | grep terraform
6030 pts/0    Sl     0:01 terraform console
```

if I then run:

```bash
$ echo "local.a" > /tmp/srv-input
```

the output file, `output.txt` is empty.

```bash
$ cat output.txt
$ 
```

If I run:

```bash
$ echo "local.c" > /tmp/srv-input # invalid input
```

the output file, `output.txt` contains the (expected) error:

    $ cat output.txt
    ╷
    │ Error: Reference to undeclared local value
    │
    │   on <console-input> line 1:
    │   (source code not available)
    │
    │ A local value with the name "c" has not been declared. Did you mean "a"?
    ╵
    
    [1]+  Exit 1                  tail -f /tmp/srv-input | terraform console >> output.txt 2>&1

----------

Why is only the stderr being redirected to the log file, but not stdout? I'm open to alternatives too
```


### Terraform Configuration Files

```terraform
...terraform config...
```

### Debug Output

.

### Expected Behavior

.

### Actual Behavior

.

### Steps to Reproduce

.

**Comment by jbardin** [comment]
Hi @BatmanAndJoker123,

Thanks for filing the issue. The lack out output immediately after piping in a successful command should be fairly easy to explain, because stdout is buffered and not a terminal, and won't be flushed right away. I haven't looked at that code in a while, but I believe the mode of operation for taking input via a non-terminal stdin is probably to read a single command, then exit after stdin closes. Stderr is traditionally unbuffered, which is why you would see that immediately. 

The fact that Terraform isn't dealing well with multiple sequential commands in this way is probably a bug of some sort however. Putting the cli into the error path should still flush all the buffers, though fixing the issue probably would prevent it from accepting a second command in the first place.

**Comment by BatmanAndJoker123** [comment]
@jbardin 

Thank you for your comment.

> The lack out output immediately after piping in a successful command should be fairly easy to explain, because stdout is buffered and not a terminal, and won't be flushed right away

Shouldn't there be an option to have it unbuffered? This is fairly common

**Comment by jbardin** [comment]
Terraform was never designed for to be placed in a pipeline where the output buffering really mattered, so it's just not something which has come up. If using the CLI in this manner has a compelling use case, we will definitely take this into consideration.

**Comment by jbardin** [comment]
Taking a quick look at the code, the intended behavior for the non-interactive mode is rad the command from the input, and then write the final result once. Concurrently reading multiple commands and writing a stream was not part of the design, so only one single output will be written after stdin has been closed.

**Comment by BatmanAndJoker123** [comment]
You're right; killing the `tail` command straight after the first command results in the stdout being written to the file.

What I'd like to have is `terraform console` running in the background and I would just feed stuff to it via the named pipe. I would like it so both stdout and stderr were written to the log file immediately, and for the process not to die immediately after an error.

`terraform console` is very slow to open if the state file is large, and having to close it/open it every time I want to use it is very annoying/slow.

**Comment by BatmanAndJoker123** [comment]
So to clarify:

`terraform console` is slow to start, and I don't want to have to open a `terraform console` process every single time I want to have something evaluated.

I want to have a background process listening, and whenever I need to evaluate something I just write to the named pipe and get the output immediately

**Comment by BatmanAndJoker123** [comment]
By the way, I'd be more than happy to contribute to this feature
